### PR TITLE
refactor service controller

### PR DESCRIPTION
**Pull request by mqliang**
This PR refactor service controller:
- use informer to list&watch node, it's much more readable
- make `serviceQueue` as a field of ServiceController
- move all the helper function to servicecontroller_utils.go, there are too many helper functions.

The code is much more readable after refactoring.

### CODE REVIEW

**Comment by mqliang** [comment]
@k8s-bot unit test this

**Comment by mqliang** [comment]
@smarterclayton @erictune

**Comment by derekwaynecarr** [comment]
I am not the best expert in the service controller.  unassigning for someone else to review.

**Comment by erictune** [comment]
@kubernetes/goog-csi team to find a reviewer.

**Comment by lavalamp** [comment]
Drive by code review: @mqliang for a next step, it looks like this code could make use of the `workqueue` I made for the endpoints controller, if we wanted to parallelize it.

(Also, I think specific controllers are probably @kubernetes/goog-control-plane, not CSI)

**Comment by mqliang** [comment]
> for a next step, it looks like this code could make use of the workqueue I made for the endpoints controller, if we wanted to parallelize it

Frankly speaking, I also think about it, but it may be a little challenging. Since when we reconcile a DELETE event, we need the deleted service to delete the according LoadBalancer,see: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/servicecontroller.go#L259

But if we use `Informer` to list&watch Service and put all the Add&Update&Delete events(key of the service) to a `workqueue`, and run several routine each pop a key from the `workqueue` and process, when processing a DELETE event, the service object has been remove from the underlying Store.

**Comment by mqliang** [comment]
@lavalamp 

Reworked(use `workqueue` to sync service). PTAL

Since we need process the DELETE event, I put `*cache.Delta` to the work queue, instead of the key of service.

**Comment by mqliang** [comment]
@a-robinson

File: pkg/controller/service/servicecontroller.go [hunk 1]
```diff
@@ -169,27 +182,33 @@ func (s *ServiceController) init() error {
...
-		err, shouldRetry := s.processDelta(delta)
-		if shouldRetry {
-			// Add the failed service back to the queue so we'll retry it.
-			glog.Errorf("Failed to process service delta. Retrying: %v", err)
-			time.Sleep(processingRetryInterval)
-			serviceQueue.AddIfNotPresent(deltas)
-		} else if err != nil {
-			util.HandleError(fmt.Errorf("Failed to process service delta. Not retrying: %v", err))
-		}
+		func() {
```
Comment (mikedanese) [inline pkg/controller/service/servicecontroller.go]: Is there a reason to wrap this in a func?

Comment (mikedanese) [inline pkg/controller/service/servicecontroller.go]: Nevermind, I see the defer

File: pkg/controller/service/servicecontroller.go [hunk 2]
```diff
@@ -169,27 +182,33 @@ func (s *ServiceController) init() error {
...
+			}
+			defer s.serviceQueue.Done(item)
+
+			delta, ok := item.(*cache.Delta)
+			if delta == nil {
+				glog.Errorf("Received nil delta from watcher queue.")
+				return
+			}
+			if !ok {
+				glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", item)
```
Comment (mikedanese) [inline pkg/controller/service/servicecontroller.go]: SHould we return on this?

**Comment by mikedanese** [comment]
@k8s-bot  ok to test

**Comment by mqliang** [comment]
@mikedanese e2e test failed. Could you please kindly paste the output. And just curious, is it feasible grant developers access to the e2e test result?

**Comment by smarterclayton** [comment]
You should be able to click the build log link in the comment.

**Comment by mqliang** [comment]
> You should be able to click the build log link in the comment.
I click the `Details` link, and login use my GitHub account(the primary email on my GitHub homepage)
image

But unfortunately, it shows `access not possible`

image

No idea why I could not access this.

**Comment by mikedanese** [comment]
@mqliang  Are you able to access this build log? 

https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/58ddc491d7faf09c472ca287d6cf8cb081412f2f/kubernetes-pull-build-test-e2e-gce/25191/build-log.txt

This is from Build Log point in the comment https://github.com/kubernetes/kubernetes/pull/19173#issuecomment-174944951

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by fejta** [comment]
This PR has had no meaningful activity for multiple months. If it is still valid please rebase, push a new commit and reopen the PR. Thanks!

**Comment by quinton-hoole** [comment]
This looks like a useful refactoring that I think we should try a little harder to get in.  Re-opening.
@mqliang My apologies for the delay.  Can you please refactor, and I'll find a reviewer for you (or do it myself).  Thanks!

**Pull request closed without merging**